polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp

```diff
@@ -1312,106 +1282,60 @@ class RaiseSetNDRange
-                  .Case<sycl::NdRangeType>(
-                      [=](auto) { return annotationStr == "nd_range"; })
-                  .Case<sycl::IDType>(
-                      [=](auto) { return annotationStr == "offset"; })
-                  .Case<sycl::RangeType>(
-                      [=](auto) { return annotationStr == "range"; })
-                  .Default(false)) &&
-             "Invalid constructor type");
-
-      constructors.push_back(constructor);
-      annotations.push_back(annotation);
+      ndRangePassed = annotationStr == "nd_range";
+
+      arguments.emplace_back(annotation, annotation.getVal());
 
       return WalkResult::advance();
     };
-    if (op.walk(getConstructors).wasInterrupted() || constructors.empty())
+    if (op.walk(getArguments).wasInterrupted() || arguments.empty())
       return failure();
 
+    assert(
+        (arguments.size() == 1 || (arguments.size() == 2 && !ndRangePassed)) &&
+        "Expecting nd-range or global size+[offset] combination");
+
     // We will introduce the operation after the last annotation
     OpBuilder::InsertionGuard ig(rewriter);
-    rewriter.setInsertionPointAfter(annotations.back());
-    Location loc = annotations.back().getLoc();
+    LLVM::VarAnnotation lastAnnotation = arguments.back().getAnnotation();
+    rewriter.setInsertionPointAfter(lastAnnotation);
+    Location loc = lastAnnotation.getLoc();
 
     // Remove annotations, no longer needed
     // Also prevents infinite recursion
     rewriter.updateRootInPlace(op, [&] {
-      for (LLVM::VarAnnotation annotation : annotations)
-        rewriter.eraseOp(annotation);
+      for (const Argument &arg : arguments)
+        rewriter.eraseOp(arg.getAnnotation());
     });
 
-    assert((constructors.size() < 3 &&
-            isa<sycl::NdRangeType, sycl::RangeType>(
-                constructors.front().getType().getValue()) &&
-            (constructors.size() == 1 ||
-             (isa<sycl::RangeType>(constructors.front().getType().getValue()) &&
-              isa<sycl::IDType>(constructors.back().getType().getValue())))) &&
-           "Expecting nd-range or global size+[offset] combination");
-
-    // TODO: nd_range constructor receives pointers to structs, so we are not
-    // able to handle this yet.
-    if (isa<sycl::NdRangeType>(constructors.front().getType().getValue()))
-      return failure();
-
-    SmallVector<Value> arguments = getArguments(op, constructors, rewriter);
-
     // Finally insert sycl.host.handler.set_nd_range after the last annotation
     rewriter.updateRootInPlace(op, [&] {
-      rewriter.create<sycl::SYCLHostHandlerSetNDRange>(loc, handler, arguments);
+      rewriter.create<sycl::SYCLHostHandlerSetNDRange>(
+          loc, handler, arguments[0].getValue(),
+          // Pass empty value if offset is not found
+          arguments.size() == 2 ? arguments[1].getValue() : Value(),
+          // Set nd_range_passed attribute if so
```

Comment:
Should we define custom builders for the operation to make it possible to pass a `bool` instead of having to construct the attribute here?

Reply:
Done!